### PR TITLE
Enforce gradients are accumulated in fp32

### PR DESCRIPTION
**Pull request by Tcc0403**
## Testing Done

- Hardware Type: <BLANK>
- [ ] run `make test` to ensure correctness
- [ ] run `make checkstyle` to ensure code style
- [ ] run `make test-convergence` to ensure convergence

### CODE REVIEW

**Comment by Tcc0403** [comment]
@qingquansong @vaibhavjindal @shimizust @lancerts 
 
#512, #827 show that training with liger flce may lead to worse performance or unstable/divergent training.
 
In mixed precision training, accumulating intermediate results in fp32 is a common trick to prevent underflow/overflow issue. 
image

However, in current liger flce, gradients of weight for linear layer are accumlated in bf16 chunk by chunk.

Allocating fp32 buffer instead of the original precision for accumlations may solve the issue, but at the cost of larger memory footprints. Besides, the benchmarking shows that the changes can also improve the speed in larger batch_size * seq_len.

There are two default values for `accum_dtype`. I'm not sure which one is more preferred:
1. `None` : Accumulating in original precision
	- matching the current behavior of flce, better backward compatibility
	- still suffering from mentioned issues
2. `torch.float32` : Accmulating in fp32
  	- more memory required
	- more stable training
	- causing potential oom for users if upgrading liger kernel without knowing the change


or any alternative solutions are welcome.

I'll cleanup this PR once we decide the interface.

File: src/liger_kernel/ops/fused_linear_cross_entropy.py [hunk 1]
```diff
@@ -124,16 +131,7 @@ def fused_linear_cross_entropy_forward(
...
-                input=grad_weight,
-                mat1=logits_chunk.t().to(
-                    _input_chunk.dtype
-                ),  # In an autocast scenario without bias, differing logits_chunk data types will cause an addmm operation error.
-                mat2=_input_chunk,
-                out=grad_weight,
-                alpha=1.0,
-                beta=1.0,
-            )
+            grad_weight += torch.mm(grad_logits_chunk.t(), _input_chunk).float()
```
Comment (Tcc0403) [inline src/liger_kernel/ops/fused_linear_cross_entropy.py:134]: Another option is performing matmul in fp32 to maintain precision, but slowing down the speed.

**Comment by shimizust** [comment]
@Tcc0403 Thanks for raising, I think the current accumulation method should be the default, as it works for most use cases.  Folks can have the option of switching to fp32 accumulation
